OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnSubterraneanLayer.cs

```diff
@@ -0,0 +1,85 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Grants Condition on subterranean layer. Also plays transition audio-visuals.")]
+	public class GrantConditionOnSubterraneanLayerInfo : GrantConditionOnLayerInfo
+	{
+		[Desc("Dig animation image to play when transitioning.")]
+		public readonly string SubterraneanTransitionImage = null;
+
+		[SequenceReference("SubterraneanTransitionImage")]
+		[Desc("Dig animation sequence to play when transitioning.")]
+		public readonly string SubterraneanTransitionSequence = null;
+
+		[PaletteReference]
+		public readonly string SubterraneanTransitionPalette = "effect";
+
+		[Desc("Dig sound to play when transitioning.")]
+		public readonly string SubterraneanTransitionSound = null;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionOnSubterraneanLayer(init.Self, this); }
+	}
+
+	public class GrantConditionOnSubterraneanLayer : GrantConditionOnLayer<GrantConditionOnSubterraneanLayerInfo>, INotifyVisualPositionChanged
+	{
+		WDist transitionDepth;
+
+		public GrantConditionOnSubterraneanLayer(Actor self, GrantConditionOnSubterraneanLayerInfo info)
+			: base(self, info, CustomMovementLayerType.Subterranean) { }
+
+		protected override void Created(Actor self)
+		{
+			var mobileInfo = self.Info.TraitInfo<MobileInfo>();
+			var li = (SubterraneanLocomotorInfo)mobileInfo.LocomotorInfo;
```

Comment:
Can we please add a `RulesetLoaded` check on the info to throw a `YamlException` if `mobileInfo.LocomotorInfo` isn't a `SubterraneanLocomotorInfo`?

Reply:
Done (on *JumpjetLayer as well).